src/rdpc/analysesProcessor.ts

```diff
@@ -0,0 +1,281 @@
+    const donorWithLatestRun = analysis.donors.reduce<DonorDocMap>(
+      (_acc, donor) => {
+        const inputAnalysesMap = _(analysis.runs)
+          .groupBy((run) => {
+            return (
+              HashCode(run.inputAnalyses[0].analysisId) +
+              HashCode(run.inputAnalyses[1]?.analysisId)
+            );
+          })
+          .value() as InputAnalysesRunMap;
+
+        Object.entries(inputAnalysesMap).forEach(([inputId, runs]) => {
+          const latestRun = getLatestRun(runs);
+          const run = latestRun === undefined ? [] : [latestRun];
+          const existingMap = _acc[donor.donorId];
+          _acc[donor.donorId] = {
+            ...existingMap,
+            [inputId]: run,
+          };
+        });
+
+        return _acc;
+      },
+      {}
+    );
+
+    // merge donor-inputAnalyses-run map by donorId, in case same donors appear under multiple analyses
+    Object.entries(donorWithLatestRun).forEach(
+      ([donorId, inputAnalysesMap]) => {
+        const existingMap = acc[donorId] ? acc[donorId] : {};
+        acc[donorId] = {
+          ...inputAnalysesMap,
+          ...existingMap,
+        };
+      }
+    );
+
+    return acc;
+  }, {});
+
+  return result;
+};
+
+const getLatestRun = (runs: Run[]): Run | undefined => {
+  return _(runs)
+    .sortBy(
+      (run) =>
+        (({ COMPLETE: 1, RUNNING: 2, EXECUTOR_ERROR: 3 } as {
+          [k: string]: number;
+        })[run.state])
+    )
+    .head();
+};
+
+export const mergeDonorMaps = (
+  mergedMap: DonorDocMap,
+  toMerge: DonorDocMap
+): DonorDocMap => {
+  Object.entries(toMerge).forEach(([donorId, inputAnalysesMap]) => {
+    const existingMap = mergedMap[donorId] ? mergedMap[donorId] : {};
+    const mergedInputAnalysesMap = {
+      ...existingMap,
+      ...inputAnalysesMap,
+    };
+    mergedMap[donorId] = { ...mergedInputAnalysesMap };
+  });
+
+  return mergedMap;
+};
+
+export const getAllMergedDonor = async (
+  studyId: string,
+  url: string,
+  config?: {
+    chunkSize?: number;
+    state?: StreamState;
+  }
+): Promise<DonorDocMap> => {
+  const stream = analysisStream(studyId, url, config);
+  let mergedDonorsWithAlignmentRuns: DonorDocMap = {};
+
+  for await (const page of stream) {
+    logger.info(`Streaming ${page.length} sequencing experiment analyses...`);
+    const donorPerPage = toDonorCentric(page);
+    mergedDonorsWithAlignmentRuns = mergeDonorMaps(
+      mergedDonorsWithAlignmentRuns,
+      donorPerPage
+    );
+  }
+  return mergedDonorsWithAlignmentRuns;
+};
+
+export const donorStateMap = (donorMap: DonorDocMap): DonorRunStateMap => {
+  let result: DonorRunStateMap = {};
+  Object.entries(donorMap).forEach(([donorId, map]) => {
+    Object.entries(map).forEach(([inputAnalysesId, runs]) => {
+      runs.forEach((run) => {
+        if (run.state === "COMPLETE") {
+          if (result[donorId]) {
+            result[donorId].alignmentsCompleted += 1;
```

Comment:
how do we know that it's alignments here?

Reply:
because this pr is specifically for alignment runs, in the next pr I will add logic to handle vc runs.